Extract Terraform tooling from Eve

gdemonet opened the pull request:
We extract our Terraform definition files from the multiple-nodes Eve worker into a user-facing configurable module. This will help both end-users wanting to try MetalK8s on OpenStack and wanting to setup various test environments. We also make sure the CI uses this new module, to gain confidence on its quality.

In the process, some meaningful changes were added to previous Terraform
definitions:
- ability to spawn private networks with Port Security disabled (so we can use them
  without IP-IP encapsulation)
- ability to use existing subnets instead of creating those private networks
- ability to provision a MetalK8s ISO, either a local one (copied with SCP) or a remote one (fetched on the Bootstrap directly)
- ability to install the Bootstrap node if an ISO was provisioned
- ability to provision Volumes for Prometheus and Alertmanager (available since 2.0 but not going to work unless the installed cluster is 2.4 or higher)
- Bastion will have a forward HTTP proxy (`squid`) setup if we use private networks (because in that case, all nodes except Bastion will be offline), and nodes will have their Yum core repositories configured to use this proxy
- Bastion can be set up to have required dependencies for running a test suite, including cloning `scality/metalk8s` on a specific branch (this is most useful for the CI and developers)
- the script that was used to create `BootstrapConfiguration` now also takes care of pre-seeding the Salt Minion ID, and provisioning of an SSH identity for the Bootstrap Salt Master
- SSH identities of Bastion and Bootstrap are properly generated on their respective hosts, and then public keys are shared to the other cluster members

We also discovered some hidden bugs:
- Some tests still relied on having the (Bootstrap) Minion ID equal to the hostname (which is the default, though not always the case)
- SaltAPI tests used to pick _any_ control-plane IP, which may not work if a VIP is configured
- The `bootstrap.sh` script was supposed to work from a non-standard mountpoint (e.g. `/var/tmp/metalk8s`), however it failed when trying to install packages from the standard mountpoint (`/srv/scality/metalk8s-<version>`) - we add a test in that script to run `state.sls metalk8s.archives.mounted` if necessary

See `contrib/terraform/README.md` for how to use this new module, which includes a table of all accepted input variables.

NicolasT commented:
Note: #2106 removes the use of VIPs altogether (until we have a `LoadBalancer` in place, of course).

gdemonet commented:
> Note: #2106 removes the use of VIPs altogether (until we have a `LoadBalancer` in place, of course).

Good point. I will remove the `apiServer` section in `deploy/scripts/bootstrap-config.sh` then, as soon as your PR lands.

gdemonet committed:
ci,contrib/tf: Extract Terraform tooling from Eve

We move the Terraform states defined in
`eve/workers/openstack-multiple-nodes/terraform` to
`contrib/terraform` for easier access by human users.

gdemonet committed:
contrib/tf: Remove unused `get-ip-leases` script

This script was used to assign IP addresses to the interfaces connected
on private networks, but we don't spawn private networks anymore, and
we won't use this approach when we re-introduce them.

gdemonet committed:
contrib/tf: Split networks and security groups

For better readability, we start re-organizing the files in our root
module. In this commit, we split the `network` definition into one
section dedicated to networks, subnets, etc. and another dedicated to
security groups.
We also extract user-facing variables into a single `variables.tf` file,
which helps readers find out all customization entrypoints for this
module in a single file.

gdemonet committed:
contrib/tf: Group re-usable locals and resources

Some variables were extracted and stored in `variables.tf`. Some re-used
locals and resources are grouped in `main.tf`, which also includes
version constraints.

The `main.tf` / `variables.tf` / `outputs.tf` layout is a best-practice
for Terraform modules. Our `main.tf` can however not include all the
logic for this module, so we still maintain additional `.tf` files (e.g.
`networks.tf`).

gdemonet committed:
contrib/tf: Split nodes deployment

We split the deployment of Bastion and Bootstrap from `nodes.tf` into
their respective files. Variables and outputs are extracted as well.

Note that the configuration of OpenStack flavours is extended to allow
finer control between the different kinds of nodes (we also increase the
default VM size for Bootstrap and Nodes to reduce the chance of
congestion).

gdemonet committed:
contrib/tf: Generate and share SSH identities

Both Bastion and Bootstrap need to have an SSH identity that other nodes
in the cluster need to authorize.
For Bastion, it allows the execution of tests based on `testinfra`.
For Bootstrap, it allows it to use Salt-SSH for cluster expansions.

gdemonet committed:
contrib/tf: Split generation of SSH config files

It may be useful to have the ability to only generate one of those.
Also, we add a "destroy"-time provisioner to clean-up the local SSH
config file.

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/extract-terraform-as-module` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `feature/extract-terraform-as-module`
* Rebase `feature/extract-terraform-as-module` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/feature/extract-terraform-as-module` with contents from `w/2.3/feature/extract-terraform-as-module`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/feature/extract-terraform-as-module origin/development/2.4
 $ git merge origin/w/2.3/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/feature/extract-terraform-as-module
```

gdemonet committed:
contrib/tf: Make Bastion optional

gdemonet committed:
contrib/tf: Deploy a forward HTTP proxy on Bastion

We deploy Squid with a basic configuration, only forwarding traffic
coming from the default subnet.
This will allow Bootstrap and other nodes to selectively access online
services, such as package repositories or container registries.

We also now leave the nodes and Bootstrap online if there is no Bastion
to proxy such traffic.

gdemonet committed:
contrib/tf: Allow the use of private networks

We re-enable private networks to avoid the need for IPIP encapsulation,
since those private networks won't have PortSecurity enabled.
All VMs are attached to each enabled private network using OpenStack
ports, after the actual VM creation, which allows to keep security
groups on the default network.

Users must enable the use of these networks using the `private_networks`
Terraform variable.
Optionally, users can provide existing subnets instead of spawning a
network and subnet with default values.
Users can also choose a VIP for which to create a port in each private
subnet.

gdemonet committed:
contrib/tf: Provision a MetalK8s ISO on Bootstrap

A user can now provide a `metalk8s_iso` Terraform variable to make sure
an ISO is provisioned on the Bootstrap node.
One can use `mode="local"` to `scp` the ISO from one's host, or
`mode="remote"` to let the Bootstrap fetch the ISO from a URL (using
Bastion as proxy if necessary).

gdemonet committed:
contrib/tf: Configure and install Bootstrap

Users can enable automatic configuration and installation of the
Bootstrap node if with `-var=metalk8s_bootstrap=true`. This assumes an
ISO was provisioned on the Bootstrap node.

Note that the `bootstrap-config.sh` script was rewritten in Python and
extended to include:
- pre-seeding of Salt minion ID
- preparing Salt master's SSH identity
- keepalived configuration (if a VIP was configured)
- network configuration detected from interfaces

gdemonet committed:
contrib/tf: Provision Volumes on the Bootstrap

This provisioner, enabled with `-var=metak8s_provision_volumes=true`,
creates SparseLoopDevice-backed Volumes for Prometheus and AlertManager.

gdemonet committed:
contrib/tf: Configure Bastion to run E2E tests

We add an optional provisioner to setup Bastion with the sources from a
chosen branch (or the default one) and the required dependencies to run
tests (using `tox`).
This will help CI and developers alike, as one can specify an arbitrary
branch name.

gdemonet committed:
contrib/tf: Enable IPIP if necessary

After installing Bootstrap, if we configured the cluster to use the
default network for at least one of the two planes, we enable IPIP
encapsulation, as default network will have PortSecurity enabled.

gdemonet committed:
contrib/tf: Add README

This gives a few examples and pointers for users to dive into the module
and start using it. Configuration options should be documented better in
`variables.tf` for ease of discovery as well as maintenance.

bert-e commented:
# History mismatch



Merge commit #afcefe464772b6de1deae6c435cfbad85d0d2166 on the integration branch
`w/2.1/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #add1975687ccdafba88db947544a98d2a3f868a8 on the integration branch
`w/2.1/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet committed:
contrib/tf: Allow specifying a VIP's number

Instead of explicitly picking a VIP's value, we let users set it to a
number defining the position of the VIP within the subnet range
configured.

Note that we bump the required version of Terraform to 0.12.7 and
higher, to be able to use the `regexall` function. In the CI, we switch
to the latest: 0.12.20

alexandre-allard commented on contrib/terraform/bastion.tf:
```diff
@@ -0,0 +1,232 @@
+}
+
+locals {
+  bastion_ip = openstack_compute_instance_v2.bastion[0].access_ip_v4
```
It'll throw an error if `bastion_enabled` is set to false.

gdemonet committed:
scripts: Ensure archive is mounted before bootstrap

In cases where the initial archive is not mounted at
/srv/scality/metalk8s-<version>, bootstrap will fail to install packages
due to files not being found at the expected path. The required state to
mount archives at their expected path is only called after, in
`metalk8s.roles.bootstrap`.
We add a simple method in the bootstrap script to execute
`metalk8s.archives.mounted` before actually installing anything.

tcarmet reviewed (commented):
Very cool PR. I actually played a little bit with the existent terraform and added a rule to setup the nodes config. It helped me to speed up a little bit my development as I was scratching up and down metal clusters.

How do you feel like adding a small terraform template like the following:
```
%{ for node in nodes ~}
---
apiVersion: v1
kind: Node
metadata:
  annotations:
    metalk8s.scality.com/ssh-host: ${node.ip}
    metalk8s.scality.com/ssh-key-path: /etc/metalk8s/pki/salt-bootstrap
    metalk8s.scality.com/ssh-port: "22"
    metalk8s.scality.com/ssh-sudo: "true"
    metalk8s.scality.com/ssh-user: centos
    node.alpha.kubernetes.io/ttl: "0"
  labels:
    metalk8s.scality.com/version: 2.4.1
    node-role.kubernetes.io/node: ""
  name: ${node.name}
%{ endfor }
```

Does it suits the idea your trying to achieve here? or you prefer to let the user figure this one for himself?

I could happily contribute to something on top of your changes if you like.

gdemonet commented:
> Very cool PR. I actually played a little bit with the existent terraform and added a rule to setup the nodes config. It helped me to speed up a little bit my development as I was scratching up and down metal clusters.

That's the whole point, super happy you tried it :)

> Does it suits the idea your trying to achieve here? or you prefer to let the user figure this one for himself?
> 
> I could happily contribute to something on top of your changes if you like.

Indeed, it's exactly what we're looking for. Next step (and probably next PR) will include templates as you suggest, and automatic cluster expansion, so we get a fully-working, well configured cluster after `terraform apply`. If you want to tackle this topic, we welcome **any** contributions :) We can discuss it further in a dedicated issue once this PR gets merged!

gdemonet committed:
contrib/tf: Pick any VIP in subnet

It seems the attribution of port IP addresses is random on some OpenStack
deployments, and forcing a VIP doesn't play well with this behaviour.
To circumvent the issue, we simply let OpenStack choose the IP for our
us. In the future, having full control over which IPs are assigned and
why, to avoid conflicts, would also be interesting.

gdemonet committed:
ci: Remove unneeded step to install kubectl

We used to only ship `kubectl` and not install by default. This
behaviour changed, so we can remove the now useless step in Eve.

gdemonet committed:
tests/install: Fix Salt errors printing

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/feature/extract-terraform-as-module` with contents from `w/2.3/feature/extract-terraform-as-module`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/feature/extract-terraform-as-module origin/development/2.4
 $ git merge origin/w/2.3/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/feature/extract-terraform-as-module
```

bert-e commented:
# History mismatch



Merge commit #9d9d6bcd5577daa14171908a1db6cdd680446acf on the integration branch
`w/2.1/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #90f749109620ef61fb050b166f9e0c89ece32cd7 on the integration branch
`w/2.1/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/feature/extract-terraform-as-module` with contents from `w/2.3/feature/extract-terraform-as-module`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/feature/extract-terraform-as-module origin/development/2.4
 $ git merge origin/w/2.3/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/feature/extract-terraform-as-module
```

gdemonet commented:
/status

bert-e commented:
# Status





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/extract-terraform-as-module` into
`development/2.0`
* [w/2.1/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/extract-terraform-as-module) will be merged into `development/2.1` 
* [w/2.2/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/extract-terraform-as-module) will be merged into `development/2.2` 
* [w/2.3/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/feature/extract-terraform-as-module) will be merged into `development/2.3` 
* [w/2.4/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/feature/extract-terraform-as-module) will be merged into `development/2.4` 
* [w/2.5/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/feature/extract-terraform-as-module) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet committed:
tests/install: Ensure Salt Master is running first

Since we recently removed a step to setup Bastion before running tests
(which we know prepare as part of the Terraform deployment), the
expansion test sometimes run too soon and the Salt Master Pod is not
found.

We thus add a waiting step, as a `Given`.

bert-e commented:
# History mismatch



Merge commit #e9f2db0a61a102c944c9dc969e766615e3b0d180 on the integration branch
`w/2.1/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #a60df8858915eca94c97543cfa8cf0f0eba86d9b on the integration branch
`w/2.1/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/feature/extract-terraform-as-module` with contents from `feature/extract-terraform-as-module`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/feature/extract-terraform-as-module origin/development/2.1
 $ git merge origin/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/feature/extract-terraform-as-module
```

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/feature/extract-terraform-as-module` with contents from `feature/extract-terraform-as-module`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/feature/extract-terraform-as-module origin/development/2.1
 $ git merge origin/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/feature/extract-terraform-as-module
```

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/feature/extract-terraform-as-module` with contents from `feature/extract-terraform-as-module`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/feature/extract-terraform-as-module origin/development/2.1
 $ git merge origin/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/feature/extract-terraform-as-module
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/feature/extract-terraform-as-module` with contents from `w/2.3/feature/extract-terraform-as-module`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/feature/extract-terraform-as-module origin/development/2.4
 $ git merge origin/w/2.3/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/feature/extract-terraform-as-module
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/extract-terraform-as-module` into
`development/2.0`
* [w/2.1/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/extract-terraform-as-module) will be merged into `development/2.1` 
* [w/2.2/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/extract-terraform-as-module) will be merged into `development/2.2` 
* [w/2.3/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/feature/extract-terraform-as-module) will be merged into `development/2.3` 
* [w/2.4/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/feature/extract-terraform-as-module) will be merged into `development/2.4` 
* [w/2.5/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/feature/extract-terraform-as-module) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #b06fff49536bdf85b747fbeb359072ed7ff14fe8 on the integration branch
`w/2.1/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/feature/extract-terraform-as-module` with contents from `feature/extract-terraform-as-module`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/feature/extract-terraform-as-module origin/development/2.1
 $ git merge origin/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/feature/extract-terraform-as-module
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/feature/extract-terraform-as-module` with contents from `w/2.3/feature/extract-terraform-as-module`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/feature/extract-terraform-as-module origin/development/2.4
 $ git merge origin/w/2.3/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/feature/extract-terraform-as-module
```

gdemonet committed:
contrib/terraform: Get rid of unused VIPs

Since the introduction of the local proxy for kube-apiserver, providing
a VIP for `keepalived` is not supported anymore. We adjust the Terraform
modules and scripts to account for this change.

gdemonet committed:
contrib/terraform: Add support for RHEL

In 2.4, we have an Eve stage for testing a simple installation on RHEL
7. This used to have its own Terraform definition, but it makes sense to
simply add support for this slight difference in the shared Terraform
module.

This implies specific provisioners to manage RHSM, and a new variable to
choose which OS to user for worker machines (Bastion will remain on
CentOS, for simplicity).

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/extract-terraform-as-module` into
`development/2.0`
* [w/2.1/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/feature/extract-terraform-as-module) will be merged into `development/2.1` 
* [w/2.2/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/feature/extract-terraform-as-module) will be merged into `development/2.2` 
* [w/2.3/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/feature/extract-terraform-as-module) will be merged into `development/2.3` 
* [w/2.4/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/feature/extract-terraform-as-module) will be merged into `development/2.4` 
* [w/2.5/feature/extract-terraform-as-module](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/feature/extract-terraform-as-module) will be merged into `development/2.5` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #734db2a74f9e39aa85996ff5ab5dba3af709b4d7 on the integration branch
`w/2.2/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.2`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #5b1e79159a1dfbfcef632071e36838d293b3273b on the integration branch
`w/2.3/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.3`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #26f47d131a363605f1af5993568b966751e8ba7c on the integration branch
`w/2.5/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

NicolasT commented on contrib/terraform/.gitignore:
```diff
@@ -0,0 +1,4 @@
+.terraform/
+*.tfvars
```
One thing we may want to do:

Make the TF templates completely ignorant of the particular OpenStack environment they're deployed to (i.e., no references to anything Scality Cloud specific), and ship a `scality-cloud.tfvars` file which provides the 'defaults' for our particular cloud environment (i.e., networks, machine types, VM images,...), and document how to deploy in SC (by using `-var-file` when invoking `terraform apply`).

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+# MetalK8s deployment Terraform tooling
+
+This set of [Terraform][1] modules is provided as convenience tooling to deploy
+a fully functional MetalK8s cluster. It currently only supports OpenStack,
```
```suggestion
a fully functional [MetalK8s](https://github.com/Scality/metalk8s) cluster. It currently only supports OpenStack,
```
or whatever the right Markdown formatting is.

NicolasT commented on contrib/terraform/README.md:
````diff
@@ -0,0 +1,79 @@
+```
+
+After a `terraform apply`, or `terraform output`, you should see a list of
+IPs for you to connect to the spawned VMs using SSH (with user `centos`).
````
This username may depend on the image being used.

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+|------|-------------|:----:|:-----:|
+| `bastion_enabled` | Whether to spawn a Bastion | bool | `"true"` |
+| `bastion_proxy_port` | Port to use for the Bastion HTTP forward proxy \(squid\) | number | `"3128"` |
+| `bastion_setup_tests` | Whether to setup test dependencies on the Bastion | bool | `"false"` |
```
Which tests? Why would this tooling need to set up things for tests?

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+| `bastion_enabled` | Whether to spawn a Bastion | bool | `"true"` |
+| `bastion_proxy_port` | Port to use for the Bastion HTTP forward proxy \(squid\) | number | `"3128"` |
+| `bastion_setup_tests` | Whether to setup test dependencies on the Bastion | bool | `"false"` |
+| `bastion_test_branch` | Which branch to clone for running tests on the Bastion | string | `""` |
```
Is this really Terraform's role?

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+| `bastion_setup_tests` | Whether to setup test dependencies on the Bastion | bool | `"false"` |
+| `bastion_test_branch` | Which branch to clone for running tests on the Bastion | string | `""` |
+| `control_plane_subnet` | Existing subnet to use for the control-plane | string | `""` |
+| `control_plane_vip` | VIP to create for the control-plane | string | `"192.168.1.2"` |
```
The control-plane doesn't have a VIP (I mean, a network doesn't 'have' some IP address). This should be reworded a bit.

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+| `bastion_test_branch` | Which branch to clone for running tests on the Bastion | string | `""` |
+| `control_plane_subnet` | Existing subnet to use for the control-plane | string | `""` |
+| `control_plane_vip` | VIP to create for the control-plane | string | `"192.168.1.2"` |
+| `default_network` | The default network name to use for SSH access | string | `"tenantnetwork1"` |
```
`public_network`? Not sure what's so 'default' about it.

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+| `metalk8s_bootstrap` | Whether to install the Bootstrap node | bool | `"false"` |
+| `metalk8s_iso_destination` | Destination of the MetalK8s ISO on Bootstrap | string | `"/home/centos/metalk8s.iso"` |
+| `metalk8s_iso_mode` | How to provision the MetalK8s ISO on Bootstrap \(can be 'local' or 'remote'\) | string | `"local"` |
+| `metalk8s_iso_mountpoint` | Where to mount the MetalK8s ISO on Bootstrap before installation | string | `"/var/tmp/metalk8s"` |
```
Mounts go under `/mnt`.

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+| `metalk8s_iso_mountpoint` | Where to mount the MetalK8s ISO on Bootstrap before installation | string | `"/var/tmp/metalk8s"` |
+| `metalk8s_iso_source` | Source URI for the MetalK8s ISO \(can be a local path or public URL\) | string | `""` |
+| `metalk8s_provision_volumes` | Whether to provision Volumes for Prometheus and AlertManager | bool | `"false"` |
+| `nodes_count` | Number of nodes to spawn in addition to Bootstrap and Bastion | string | `"1"` |
```
What kind of nodes does this refer to?

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+| `metalk8s_iso_source` | Source URI for the MetalK8s ISO \(can be a local path or public URL\) | string | `""` |
+| `metalk8s_provision_volumes` | Whether to provision Volumes for Prometheus and AlertManager | bool | `"false"` |
+| `nodes_count` | Number of nodes to spawn in addition to Bootstrap and Bastion | string | `"1"` |
+| `openstack_flavours` |  | map | `{ bastion = "m1.medium", bootstrap = "m1.large", nodes = = "m1.large" }` |
```
There's a duplicate `= =` in the `nodes` part.

NicolasT replied:
As referred to above, I think this should go in some SC-specific vars file instead of being a (supposedly cross-platform) default.

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+| `openstack_image_name` | OpenStack configuration | string | `"CentOS-7-x86_64-GenericCloud-1809.qcow2"` |
+| `openstack_link_local_ip` | Metadata service used by cloud-init (see https://docs.openstack.org/nova/latest/user/metadata-service.html) | string | `"169.254.169.254"` |
+| `private_networks` | Which private networks to use \(or spawn\) | list | `[]` |
+| `ssh_key_pair` | The SSH key pair to use for setting up OpenStack VMs | map | `{ private_key = "~/.ssh/terraform", public_key  = "~/.ssh/terraform.pub" }` |
```
Can't we generate a keypair through the templates?

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+| `openstack_link_local_ip` | Metadata service used by cloud-init (see https://docs.openstack.org/nova/latest/user/metadata-service.html) | string | `"169.254.169.254"` |
+| `private_networks` | Which private networks to use \(or spawn\) | list | `[]` |
+| `ssh_key_pair` | The SSH key pair to use for setting up OpenStack VMs | map | `{ private_key = "~/.ssh/terraform", public_key  = "~/.ssh/terraform.pub" }` |
+| `worker_uuid` | Use in CI to link a deployment to its worker | string | `""` |
```
I'm not sure what this means, and I'm not sure this needs to be in the 'public' ('end-user') docs.

NicolasT commented on contrib/terraform/README.md:
```diff
@@ -0,0 +1,79 @@
+| `ssh_key_pair` | The SSH key pair to use for setting up OpenStack VMs | map | `{ private_key = "~/.ssh/terraform", public_key  = "~/.ssh/terraform.pub" }` |
+| `worker_uuid` | Use in CI to link a deployment to its worker | string | `""` |
+| `workload_plane_subnet` | Existing subnet to use for the workload-plane | string | `""` |
+| `workload_plane_vip` | VIP to create for the workload-plane | string | `""` |
```
Same as above (even less confused by this one, actually).

NicolasT reviewed (changes requested):
Intermediate publishing of review comments.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

NicolasT commented on contrib/terraform/bastion.tf:
```diff
@@ -0,0 +1,240 @@
+  count = local.bastion.enabled ? 1 : 0
+
+  name        = "${local.prefix}-bastion"
+  image_name  = var.openstack_image_name
```
Maybe these should be kept in a map as well (similar to how machine flavours are mapped).

NicolasT commented on contrib/terraform/bastion.tf:
```diff
@@ -0,0 +1,240 @@
+  key_pair    = openstack_compute_keypair_v2.local_ssh_key.name
+
+  scheduler_hints {
+    group = openstack_compute_servergroup_v2.all_machines.id
```
You're putting all VMs on the same hypervisor? :thinking: :thinking: :thinking:

NicolasT commented on contrib/terraform/bastion.tf:
```diff
@@ -0,0 +1,240 @@
+  # Generate an SSH keypair
+  provisioner "remote-exec" {
+    inline = [
+      "ssh-keygen -t rsa -b 4096 -N '' -f /home/centos/.ssh/bastion"
```
Can we instead generate an SSH key in Terraform and use this one (provision the private part on the bastion, public part on other nodes), instead of shelling out?

NicolasT commented on contrib/terraform/bastion.tf:
```diff
@@ -0,0 +1,240 @@
+  connection {
+    host        = self.access_ip_v4
+    type        = "ssh"
+    user        = "centos"
```
Since VM images are parametrized, this is not necessarily true.

NicolasT commented on contrib/terraform/bastion.tf:
```diff
@@ -0,0 +1,240 @@
+  # Provision scripts for remote-execution
+  provisioner "file" {
+    source      = "${path.root}/scripts"
+    destination = "/home/centos/scripts"
```
Can we use some temp path, not tied to the `centos` user instead?

NicolasT commented on contrib/terraform/bastion.tf:
```diff
@@ -0,0 +1,240 @@
+  connection {
+    host        = openstack_compute_instance_v2.bastion[0].access_ip_v4
+    type        = "ssh"
+    user        = "centos"
```
Same as above (and likely applies in other places as well).

NicolasT commented on contrib/terraform/bastion.tf:
```diff
@@ -0,0 +1,240 @@
+        ? [openstack_networking_port_v2.workload_plane_bastion[0].mac_address]
+        : [],
+      ) :
+      "sudo bash scripts/network-iface-config.sh ${mac_address}"
```
I'm surprised we can't get this done using an injected `cloud-init` config instead.

NicolasT commented on contrib/terraform/bastion.tf:
```diff
@@ -0,0 +1,240 @@
+#        maybe simply suggest using `sshuttle` with the Bastion?
+# NOTE: after this is provisioned, one can run E2E tests with `tox` from the
+#       cloned repository on the Bastion
+resource "null_resource" "setup_tests_bastion" {
```
I seriously doubt this should be part of what TF does.

NicolasT commented on contrib/terraform/bootstrap.tf:
```diff
@@ -0,0 +1,174 @@
+
+  network {
+    access_network = true
+    name           = data.openstack_networking_network_v2.default_network.name
```
Is the `bootstrap` node required to be attached to this network?

NicolasT commented on contrib/terraform/bootstrap.tf:
```diff
@@ -0,0 +1,174 @@
+  # Generate an SSH keypair
+  provisioner "remote-exec" {
+    inline = [
+      "ssh-keygen -t rsa -b 4096 -N '' -f /home/centos/.ssh/bootstrap"
```
Same as above and such.

NicolasT commented on contrib/terraform/bootstrap.tf:
```diff
@@ -0,0 +1,174 @@
+        ? [openstack_networking_port_v2.workload_plane_bootstrap[0].mac_address]
+        : [],
+      ) :
+      "sudo bash scripts/network-iface-config.sh ${mac_address}"
```
How are these files made available on the target system? :thinking:

NicolasT commented on contrib/terraform/main.tf:
```diff
@@ -0,0 +1,35 @@
+# Versions
+terraform {
+  required_version = ">= 0.12.7"
```
Shouldn't this be a `~>` as well?

NicolasT commented on contrib/terraform/metalk8s.tf:
```diff
@@ -0,0 +1,173 @@
+  # Prepare for bootstrap installation
+  provisioner "remote-exec" {
+    inline = [
+      join(" ", compact([
```
Not sure (didn't see the scripts yet), but it feels like using the TF `template_file` module to generate a 'configuration file' somehow, then either sourcing it in `prepare-bootstrap.sh` (or not needing `prepare-bootstrap.sh` at all depending on what it does?) could be more maintainable.

NicolasT commented on contrib/terraform/metalk8s.tf:
```diff
@@ -0,0 +1,173 @@
+resource "null_resource" "enable_ipip" {
+  # If one of the networks used by Metal is the default network, PortSecurity
+  # will still be enabled, so we need IPIP encapsulation
+  count = var.metalk8s_bootstrap && length(var.private_networks) < 2 ? 1 : 0
```
This is a heuristic based on how Scality Cloud is designed. Should instead be configuration, IMHO.

NicolasT commented on contrib/terraform/networks.tf:
```diff
@@ -0,0 +1,177 @@
+    name    = "${local.prefix}-control-plane",
+    # Hard-coded defaults
+    iface = "eth1",
+    cidr  = "192.168.1.0/24",
```
Why hard-coded?

NicolasT commented on contrib/terraform/networks.tf:
```diff
@@ -0,0 +1,177 @@
+    enabled = contains(var.private_networks, "control_plane")
+    name    = "${local.prefix}-control-plane",
+    # Hard-coded defaults
+    iface = "eth1",
```
I don't see this used. If I missed it, and it *is* used, would be nice to *not* rely on this in any way.

NicolasT commented on contrib/terraform/nodes.tf:
```diff
@@ -0,0 +1,175 @@
+
+  network {
+    access_network = true
+    name           = data.openstack_networking_network_v2.default_network.name
```
What if we want nodes to be offline?

NicolasT commented on contrib/terraform/nodes.tf:
```diff
@@ -0,0 +1,175 @@
+        ? [openstack_networking_port_v2.workload_plane_nodes[count.index].mac_address]
+        : [],
+      ) :
+      "sudo bash scripts/network-iface-config.sh ${mac_address}"
```
This should really be possible using `cloud-init`...

NicolasT commented on contrib/terraform/scripts/prepare-bootstrap.sh:
```diff
@@ -0,0 +1,69 @@
+# Parameters
+OUTPUT_FILE="/etc/metalk8s/bootstrap.yaml"
+ARCHIVE_PATH=${ARCHIVE_PATH:-"/home/centos/metalk8s.iso"}
+CONTROL_PLANE_NETWORK=${CP_NET:-"10.0.0.0/8"}
```
This 'default' doesn't match the 'default' in TF. As such, not sure any 'default' *can* actually make sense here. If the value is required input, then treat it as such (same applies to workload plane of course).

NicolasT commented on contrib/terraform/ssh.tf:
```diff
@@ -0,0 +1,107 @@
+  ]
+
+  # Share and authorize Bastion key
+  provisioner "local-exec" {
```
I think this is where using a TF-calculated key would prove to be somewhat easier as well.

NicolasT commented on contrib/terraform/templates/squid.conf.tpl:
```diff
@@ -0,0 +1,65 @@
+# Recommended minimum Access Permission configuration:
+#
+# Deny requests to certain unsafe ports
+http_access deny !Safe_ports
```
We only need HTTP and HTTPS so let's remove all others from these 'safe' ports.

NicolasT commented on contrib/terraform/templates/squid.conf.tpl:
```diff
@@ -0,0 +1,65 @@
+#
+# Add any of your own refresh_pattern entries above these.
+#
+refresh_pattern ^ftp:           1440    20%     10080
```
Don't need these except for the last one I guess.

NicolasT commented on eve/main.yml:
```diff
@@ -534,6 +523,10 @@ stages:
             # FIXME: this makes hostnames too long
             # TF_VAR_worker_uuid: "%(prop:worker_uuid)s"
             TF_VAR_nodes_count: "2"
```
We should use a `vars-file` instead, IMHO.

NicolasT commented on scripts/bootstrap.sh.in:
```diff
@@ -211,6 +211,18 @@ configure_salt_minion_local_mode() {
+    local -r script_dir=$(dirname "$(readlink -f "${BASH_SOURCE[0]}")")
+
+    if [[ "$script_dir" != "/srv/scality/metalk8s-@@VERSION" ]]; then
+        "${SALT_CALL}" --local --state-output=mixed --retcode-passthrough \
```
Is there a reason to not always run this state?

gdemonet committed:
contrib/tf: Add triggers for resources

This allows resources to be re-created if one of the dependencies
changes.

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/2.4/feature/extract-terraform-as-module` with contents from `w/2.3/feature/extract-terraform-as-module`
and `development/2.4`.

Please resolve the conflict on **the integration branch** (`w/2.4/feature/extract-terraform-as-module`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/2.4/feature/extract-terraform-as-module
 $ git pull  # or "git reset --hard origin/w/2.4/feature/extract-terraform-as-module"
 $ git merge origin/development/2.4
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/w/2.3/feature/extract-terraform-as-module
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/feature/extract-terraform-as-module
```

bert-e commented:
# Conflict




There is a conflict between your branch `feature/extract-terraform-as-module` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`feature/extract-terraform-as-module`).

```bash
 $ git fetch
 $ git checkout origin/feature/extract-terraform-as-module
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:feature/extract-terraform-as-module
```

gdemonet committed:
tests/salt-api: Get control-plane IP from grains

We used to get the first available IP in the control-plane network,
though this may actually be the VIP, which wouldn't work.
Instead, the metalk8s:control_plane_ip grain takes care of ignoring the
VIP, so using this will always work, whether we use a VIP or not.

bert-e commented:
# History mismatch



Merge commit #db0383faf0502b8e59cb11415665e1b3ad05cd3b on the integration branch
`w/2.1/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# History mismatch



Merge commit #db0383faf0502b8e59cb11415665e1b3ad05cd3b on the integration branch
`w/2.5/feature/extract-terraform-as-module` is merging a branch which is neither the current
branch `feature/extract-terraform-as-module` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `feature/extract-terraform-as-module` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet committed:
WIP - refactor based on reviews

gdemonet committed:
WIP - more refactor based on reviews
